packages/cosmic-swingset/src/chain-main.js

```diff
@@ -439,25 +446,137 @@ export default async function main(progname, args, { env, homedir, agcc }) {
       mailboxStorage,
       clearChainSends,
       replayChainSends,
-      setActivityhash,
       bridgeOutbound: doOutboundBridge,
       vatconfig,
       argv,
       env,
       verboseBlocks: true,
       metricsProvider,
       slogSender,
+      swingStoreExport,
       swingStoreTraceFile,
       keepSnapshots,
       afterCommitCallback,
     });
 
-    savedChainSends = s.savedChainSends;
+    let shutdown;
+    ({ blockingSend, shutdown, writeSlogObject, savedChainSends } = s);
 
-    return s.blockingSend;
+    registerShutdown(async () =>
+      Promise.all([shutdown(), discardStateSyncExport()]).then(() => {}),
+    );
+
+    return blockingSend;
+  }
+
+  async function handleCosmosSnapshot(blockHeight, request, requestArgs) {
+    switch (request) {
+      case 'restore': {
+        const exportDir = requestArgs[0];
+        if (typeof exportDir !== 'string') {
+          throw Fail`Invalid exportDir argument ${String(exportDir)}`;
+        }
+        return performStateSyncImport(
+          { exportDir, stateDir: stateDBDir, blockHeight },
+          { fs: { ...fs, ...fsPromises }, pathResolve },
+        );
+      }
+      case 'initiate': {
+        !stateSyncExport ||
+          Fail`Snapshot already in progress for ${stateSyncExport.blockHeight}`;
+
+        const exportData =
+          /** @type {Required<NonNullable<typeof stateSyncExport>>} */ ({
+            blockHeight,
+          });
+        stateSyncExport = exportData;
+
+        // eslint-disable-next-line @jessie.js/no-nested-await
+        await new Promise((resolve, reject) => {
+          tmpfs.dir(
+            {
+              prefix: `agd-state-sync-${blockHeight}-`,
+              unsafeCleanup: true,
+            },
+            (err, exportDir, cleanup) => {
+              if (err) {
+                reject(err);
+              } else {
```

Comment:
Please avoid deep nesting.
```suggestion
                reject(err);
                return;
              }
              // continue without nesting...
```